include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+       */
+#ifdef DEAL_II_WITH_ADOLC
+      // Note: This value is a limitation of ADOL-C, and not something that we
+      // have control over. See test adolc/helper_tape_index_01.cc for
+      // verification that we cannot use or exceed this value. This value is
+      // defined as TBUFNUM; see
+      // https://gitlab.com/adol-c/adol-c/blob/master/ADOL-C/include/adolc/internal/usrparms.h#L34
+      static const types::tape_index max_tape_index = TBUFNUM;
+#else
+      static const types::tape_index_type max_tape_index =
+        std::numeric_limits<types::tape_index_type>::max();
+#endif // DEAL_II_WITH_ADOLC
+    }  // namespace types
+
+    /**
+     * A driver class for taped auto-differentiable numbers.
+     *
+     * It is intended that this class be specialized for the valid
+     * combinations of auto-differentiable numbers and output scalar
+     * number types.
+     *
+     * @tparam ADNumberType A type corresponding to a supported
+     *         auto-differentiable number.
+     * @tparam ScalarType A real or complex floating point number type
+     *         that is the scalar value type used for input to, and output
+     *         from, operations performed with auto-differentiable numbers.
+     * @tparam T An arbitrary type resulting from the application of
+     *         the SFINAE idiom to selectively specialize this class.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename ADNumberType, typename ScalarType, typename T = void>
+    struct TapedDrivers
+    {
+      // This dummy class definition safely supports compilation
+      // against tapeless numbers or taped number types that have
+      // not yet been implemented.
+
+      /**
+       * @name Taping
+       */
+      //@{
+
+      /**
+       * Enable the recording mode for a given tape.
+       *
+       * @param[in] tape_index The index of the tape to be written.
+       * @param[in] keep_independent_values Determines whether the numerical
+       *            values of all independent variables are recorded in the
+       *            tape buffer.
+       */
+      static void
+      enable_taping(const types::tape_index tape_index,
+                    const bool              keep_independent_values);
+
+      /**
+       * Enable the recording mode for a given tape, using the run-time
+       * chosen tape buffer sizes.
+       *
+       * @param[in] tape_index The index of the tape to be written.
+       * @param[in] keep_independent_values Determines whether the numerical
+       *            values of all independent variables are recorded in the
+       *            tape buffer.
+       * @param[in] obufsize ADOL-C operations buffer size
+       * @param[in] lbufsize ADOL-C locations buffer size
+       * @param[in] vbufsize ADOL-C value buffer size
+       * @param[in] tbufsize ADOL-C Taylor buffer size
+       */
+      static void
+      enable_taping(const types::tape_index        tape_index,
```

Comment:
I don't recall whether I commented on this already in the previous patch, but `enable/disable_` taping sounds like you make it possible. Is this what you mean? Or do you really mean that this *starts* or *stops* taping?